### PR TITLE
[reward] extend foundation bonus

### PR DESCRIPTION
**Pull request by dustinxie**

### CODE REVIEW

File: action/protocol/rewarding/protocol_test.go [hunk 1]
```diff
@@ -539,56 +497,104 @@ func TestStateCheckLegacy(t *testing.T) {
...
+		r.Equal(f1, f)
+
+		e := exempt{}
+		_, err = p.stateV1(sm, exemptKey, &e)
+		r.NoError(err)
+		r.Equal(e1, e)
+
+		for _, h := range []uint64{
+			g.GreenlandBlockHeight,
+			g.KamchatkaBlockHeight,
```
Comment (dustinxie) [inline action/protocol/rewarding/protocol_test.go:540]: refactor the test to cover migration at both Greenland and Kamchatka height

File: action/protocol/rewarding/protocol_test.go [hunk 2]
```diff
@@ -539,56 +524,104 @@ func TestStateCheckLegacy(t *testing.T) {
...
```
Comment (dustinxie) [inline action/protocol/rewarding/protocol_test.go:567]: this test is refactored to cover migration at both Greenland and Kamchatka height

File: action/protocol/rewarding/protocol.go [hunk 3]
```diff
@@ -39,34 +39,50 @@ var (
...
 // reward amount, users to donate tokens to the fund, block producers to grant them block and epoch reward and,
 // beneficiaries to claim the balance into their personal account.
 type Protocol struct {
-	keyPrefix                   []byte
-	addr                        address.Address
-	foundationBonusP2StartEpoch uint64
-	foundationBonusP2EndEpoch   uint64
+	keyPrefix []byte
+	addr      address.Address
+	cfg       genesis.Rewarding
```
Comment (dustinxie) [inline action/protocol/rewarding/protocol.go:51]: the P2 bonus extension is not done properly, we just add 2 numbers `P2Start/End` into the `Protocol{}`, business logic to determine bonus partly depends on the `admin{}` stored in the statedb, and partly depends on these 2
see comment in reward.go

File: action/protocol/rewarding/reward.go [hunk 4]
```diff
@@ -222,7 +222,13 @@ func (p *Protocol) GrantEpochReward(
 	}
 
 	// Reward additional bootstrap bonus
-	if epochNum <= a.foundationBonusLastEpoch || (epochNum >= p.foundationBonusP2StartEpoch && epochNum <= p.foundationBonusP2EndEpoch) {
+	var grantFB bool
+	if a.hasFoundationBonusExtension() {
+		grantFB = a.grantFoundationBonus(epochNum)
+	} else {
+		grantFB = epochNum <= a.foundationBonusLastEpoch || (epochNum >= p.cfg.FoundationBonusP2StartEpoch && epochNum <= p.cfg.FoundationBonusP2EndEpoch)
```
Comment (dustinxie) [inline action/protocol/rewarding/reward.go:229]: grant bonus depends on both `admin{}` stored in statedb, and `P2Start/End` in local struct

at Kamchatka height, we add the bonus Start/End epoch, so it solely depends on `admin{}` stored in statedb

File: action/protocol/rewarding/protocol.go [hunk 5]
```diff
@@ -96,6 +112,8 @@ func (p *Protocol) CreatePreStates(ctx context.Context, sm protocol.StateManager
 		return p.SetReward(ctx, sm, g.DardanellesBlockReward(), true)
 	case g.GreenlandBlockHeight:
 		return p.migrateValueGreenland(ctx, sm)
+	case g.KamchatkaBlockHeight:
+		return p.setFoundationBonusExtension(ctx, sm)
```
Comment (dustinxie) [inline action/protocol/rewarding/protocol.go:116]: at Kamchatka height, we add the bonus Start/End epoch, so the business logic now solely depends on `admin{}` stored in statedb

File: action/protocol/rewarding/protocol.go [hunk 6]
```diff
@@ -39,34 +39,50 @@ var (
...
+		keyPrefix: h[:],
+		addr:      addr,
+		cfg:       cfg,
+	}
+}
+
+// verify that foundation bonus extension epochs are in increasing order
+func validateFoundationBonusExtension(cfg genesis.Rewarding) error {
+	end := cfg.FoundationBonusLastEpoch
+	if cfg.FoundationBonusP2StartEpoch < end || cfg.FoundationBonusP2EndEpoch < cfg.FoundationBonusP2StartEpoch {
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply action/protocol/rewarding/protocol.go]: good catch, updated PR

File: action/protocol/rewarding/protocol.go [hunk 7]
```diff
@@ -148,12 +151,32 @@ func (p *Protocol) setFoundationBonusExtension(ctx context.Context, sm protocol.
...
+	adminExtension := len(a.foundationBonusExtension)
+	configExtension := len(p.cfg.FoundationBonusExtension)
+	if hasP2Extension {
+		configExtension++
+	}
+	if configExtension != adminExtension+1 {
+		return errInvalidEpoch
+	}
+
+	last := len(p.cfg.FoundationBonusExtension) - 1
```
Comment (dustinxie) [inline action/protocol/rewarding/protocol.go:177]: for future extension, only need to add the last/new entry

File: action/protocol/rewarding/protocol.go [hunk 8]
```diff
@@ -148,12 +151,32 @@ func (p *Protocol) setFoundationBonusExtension(ctx context.Context, sm protocol.
 		return err
 	}
 
-	a.foundationBonusExtension = append(a.foundationBonusExtension,
-		genesis.Period{p.cfg.FoundationBonusP2StartEpoch, p.cfg.FoundationBonusP2EndEpoch})
-	for _, v := range p.cfg.FoundationBonusExtension {
-		a.foundationBonusExtension = append(a.foundationBonusExtension,
-			genesis.Period{v.Start, v.End})
+	hasP2Extension := (p.cfg.FoundationBonusP2StartEpoch > 0 || p.cfg.FoundationBonusP2EndEpoch > 0)
+	if !a.hasFoundationBonusExtension() {
```
Comment (dustinxie) [inline action/protocol/rewarding/protocol.go:155]: first time add extension

**Review by CoderZhi — APPROVED** [review]

**Review by raullenchai — APPROVED** [review]

**Pull request merged**